### PR TITLE
xxhash 64 support

### PR DESCRIPTION
**Pull request by jsjhoubo**

### CODE REVIEW

File: table/block_fetcher.cc [hunk 1]
```diff
@@ -49,6 +49,9 @@ void BlockFetcher::CheckBlockChecksum() {
       case kxxHash:
         actual = XXH32(data, static_cast<int>(block_size_) + 1, 0);
         break;
+      case kxxHash64:
+        actual =static_cast<uint32_t> (XXH64(data, static_cast<int>(block_size_) + 1, 0));
```
Comment (siying) [inline table/block_fetcher.cc]: It looks over 80-char. Can you run "make format"?

File: table/block_based_table_builder.cc [hunk 2]
```diff
@@ -610,6 +610,16 @@ void BlockBasedTableBuilder::WriteRawBlock(const Slice& block_contents,
         EncodeFixed32(trailer_without_type, XXH32_digest(xxh));
         break;
       }
+      case kxxHash64: {
+        XXH64_state_t* const state = XXH64_createState();
+        XXH64_reset(state, 0);
+        XXH64_update(state, block_contents.data(),
+                static_cast<uint32_t>(block_contents.size()));
+        XXH64_update(state, trailer, 1);  // Extend  to cover block type
+        EncodeFixed32(trailer_without_type, static_cast<uint32_t>(XXH64_digest(state)));
```
Comment (siying) [inline table/block_based_table_builder.cc]: Just for code readable, I suggest we do an explicit instruction to take the lower bits before casting to uint32_t. Something like `& 0xffffffff`. The result should be the same though.

File: util/xxhash.cc [hunk 3]
```diff
@@ -476,4 +554,425 @@ U32 XXH32_digest (void* state_in)
     return h32;
 }
 
+/* *******************************************************************
+*  64-bit hash functions
+*********************************************************************/
+
+#if defined(_MSC_VER)     /* Visual Studio */
+#  define XXH_swap64 _byteswap_uint64
```
Comment (Cyan4973) [inline util/xxhash.cc]: I guess you actually want to define `XXH_read64()` here,
instead of `XXH_swap64()` which is re-defined later on.

**Comment by Cyan4973** [comment]
This should work, assuming the correction regarding `XXH_read64()` is completed.

I note that this patch selects a kind of "custom hybrid" methodology,
where the code of `XXH64` is extracted from upstream version,
and copy/pasted throughout `xxhash.cc`.
I suspect this is an attempt to not change the `XXH32` API, since it has evolved upstream.

While on immediate term this works,
it also makes future maintenance more difficult : 
any later change / fix from upstream will be more difficult to integrate,
as it requires some custom manual copy/paste operation
(instead of a simpler "let's update the whole file" operation).
Also, the resulting API is not compatible with upstream `libxxhash` (on the XXH32 part),
so no possibility to use dynamic linking in the future.

I would humbly suggest to stick to upstream version, for better maintenance, and bite the bullet by modifying XXH32 invocation throughout existing code (which, from a quick github search through the code base, seems to be fairly limited). Ultimately though, this kind of choice is for the RocksDB team.

File: table/block_based_table_builder.cc (hunk 2 above)
Comment (siying) [reply table/block_based_table_builder.cc]: I guess this is a question for @Cyan4973 . Does small endian and big endian machines get the same xxhash64 results? If they are, then 0xffffffff should return the same result.

Comment (jsjhoubo) [reply table/block_based_table_builder.cc]: @siying , no matter big endian or small endian, the static_cast<uint32_t> should be ok

**Comment by siying** [comment]
@Cyan4973 thank you for your comment. I think this is too much for a bootcamp task, but we'll find time to do the code cleaning up later.

File: table/block_based_table_builder.cc (hunk 2 above)
Comment (siying) [reply table/block_based_table_builder.cc]: @Cyan4973 thanks
@jsjhoubo I know that, but if `& 0xffffffff` generate the same result, I suggest we do that before casting to uint32_t. I personally don't like the idea of down casting with the value itself changed. I feel it hurts readability.

**Comment by siying** [comment]
I also started the landing process but I forgot one thing: can you mention this new feature in HISTORY.md?

File: include/rocksdb/table.h [hunk 4]
```diff
@@ -47,6 +47,7 @@ enum ChecksumType : char {
   kNoChecksum = 0x0,
   kCRC32c = 0x1,
   kxxHash = 0x2,
+  kxxHash64 = 0x3,
```
Comment (siying) [inline include/rocksdb/table.h]: Forgot to mention one thing. We should add a comment here, saying it's the lower 32 bits of xxhash64, not xxhash64.

**Pull request closed without merging**